### PR TITLE
fix(ssr): extra bookends for light dom forwarded slots @W-17363666

### LINKED ISSUE #5000: HTML bookends are not properly inserted for light DOM forwarded slots in @lwc/ssr-compiler
### Description

The `@lwc/ssr-compiler` currently appends HTML bookends whenever a light DOM `slot` element is encountered, which casuses excessive bookends to appear.

For instance:

```html
<template lwc:render-mode="light">
    <x-slot>
        <h1 slot="top">top content</h1>
        <h1 slot="bottom">bottom content</h1>
    </x-slot>
</template>
```

```html
<template lwc:render-mode="light">
    <x-leaf>
        <slot name="top" slot="bottom"></slot>
        <slot name="bottom" slot="top"></slot>
    </x-leaf>
</template>
// x-slot
```

```html
<template lwc:render-mode="light">
    <slot name="top"></slot>
    <slot name="bottom"></slot>
</template>
// x-leaf
```

The expected outcome would be :

```html
<x-light-container>
  <x-slot>
    <x-leaf>
      <!---->
      <h1>
        bottom content
      </h1>
      <!---->
      <!---->
      <h1>
        top content
      </h1>
      <!---->
    </x-leaf>
  </x-slot>
</x-light-container>
```
However, the contents render with extra bookends:

```html
  <x-light-container>
    <x-slot>
      <x-leaf>
        <!---->
       <!---->
        <h1>
          bottom content
        </h1>
        <!---->
        <!---->
       <!---->
       <!---->
        <h1>
          top content
        </h1>
        <!---->
       <!---->
      </x-leaf>
    </x-slot>
  </x-light-container>
```
Notice there's an extra set of bookends that wraps each `<h1>`, which corresponds to each of the `slot` elements.

`@lwc/engine-core` wraps each light DOM `slot` in a [`VFragment`](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/api.ts#L322) but when the `slot` is forwarded it will be [flattened](https://github.com/salesforce/lwc/blob/71b929eae0deea658323469f13350307e5067e0c/packages/%40lwc/engine-core/src/framework/rendering.ts#L694) before it's allocated to the next `slot`.

This prevents extra bookends from being generated.

### PR DESCRIPTION
**Pull request by cardoso**
## Details

Closes #5000

Fixes 4 expected failures:
- slot-forwarding/scoped-slots/index.js
- slot-forwarding/slots/light/index.js
- slot-not-at-top-level/lwcIf/light/index.js
- slot-not-at-top-level/lwcIf/shadow/index.js

Also narrows down the following two to the issue described in https://github.com/salesforce/lwc/pull/4999#discussion_r1868482997

- slot-forwarding/slots/dangling/index.js (5 to 1)
- slot-forwarding/slots/mixed/index.js (6 to 2)

## Does this pull request introduce a breaking change?

- 😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

- 🤞 No, it does not introduce an observable change.

## GUS work item

[W-17363666](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000026KTbKYAW/view)

### CODE REVIEW

File: packages/@lwc/ssr-compiler/src/compile-template/transformers/if.ts [hunk 1]
```diff
@@ -63,7 +63,7 @@ function bIfStatement(
 
     return b.ifStatement(
         expressionIrToEs(condition, cxt),
-        bBlockStatement(children, cxt, true),
+        bBlockStatement(children, cxt, !cxt.isSlotted),
```
Comment (cardoso) [inline packages/@lwc/ssr-compiler/src/compile-template/transformers/if.ts:66]: This removes 3 expected failures related to `lwc:if`, but the real solution for it in light-dom which would fix `slot-not-at-top-level/advanced/lwcIf/light/index.js` is a bit more complicated.

**Review by wjhsf — APPROVED** [review]

File: packages/@lwc/ssr-compiler/src/compile-template/transformers/if.ts (hunk 1 above)
Comment (jmsjtu) [reply packages/@lwc/ssr-compiler/src/compile-template/transformers/if.ts:66]: We might want to only do this if the `lwc:if` is at the top level of the slotted content. 

Looking at the [way](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L718) we flatten slotted content, we only do it for the top-level elements being slotted.

For example:

```html
<template lwc:render-mode="light">
    <x-slot>
        <h1 slot="top">top content</h1>
        <h1 slot="bottom">
            <template lwc:if={isTrue}>
                <span>rendered</span>
            </template>
        </h1>
    </x-slot>
</template>
```

```html
<template lwc:render-mode="light">
    <x-leaf>
        <slot name="top" slot="bottom"></slot>
        <slot name="bottom" slot="top"></slot>
    </x-leaf>
</template>
// x-slot
```

```html
<template lwc:render-mode="light">
    <slot name="top"></slot>
    <slot name="bottom"></slot>
</template>
// x-leaf
```

The bookends around the `lwc:if` will be removed:

```html
  <x-light-container>
    <x-slot>
      <x-leaf>
        <!---->
        <h1>
-         <!---->
          <span>
            rendered
          </span>
-         <!---->
        </h1>
        <!---->
        <!---->
        <h1>
          top content
...
```

We could probably do this in a follow-up PR though.

File: packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts [hunk 2]
```diff
@@ -168,6 +168,8 @@ export function getSlottedContent(
     node: IrLwcComponent | IrComponent,
     cxt: TransformerContext
 ): EsStatement[] {
+    cxt.isSlotted = true;
```
Comment (jmsjtu) [inline packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts:171]: 👍

File: packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts [hunk 3]
```diff
@@ -199,6 +201,8 @@ export function getSlottedContent(
         lightSlotContent.length > 0 || scopedSlotContent.length > 0
     );
 
+    cxt.isSlotted = false;
+
```
Comment (nolanlawson) [inline packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts:205]: What if `ctx.isSlotted` is already true on line 171? If we have two nested slotted contents in a row, then we want to reset back to true, not false.

Here is an example test case that fails due to this: 80a568e57

**Review by nolanlawson — CHANGES_REQUESTED** [review]
This LGTM and I do like `cxt.isSlotted`, but there is a small issue with nested slot content. Thanks very much for this! 🎉

File: packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts (hunk 3 above)
Comment (cardoso) [reply packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts:207]: @nolanlawson thanks for the test case! I added it and pushed a simple fix.

**Comment by cardoso** [comment]
Once this is merged I'll push the follow-up PR for `slot-not-at-top-level/advanced/lwcIf/light/index.js`

File: packages/@lwc/engine-server/src/__tests__/fixtures/slotted-content-nested/modules/x/slots/slots.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { LightningElement } from 'lwc';
+export default class extends LightningElement {
+    isFalse = false;
```
Comment (nolanlawson) [inline packages/@lwc/engine-server/src/__tests__/fixtures/slotted-content-nested/modules/x/slots/slots.js:3]:
```suggestion
```

Messed this up in my repro. 😆

File: packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts [hunk 5]
```diff
@@ -168,6 +168,10 @@ export function getSlottedContent(
     node: IrLwcComponent | IrComponent,
     cxt: TransformerContext
 ): EsStatement[] {
+    const isNested = cxt.isSlotted;
```
Comment (nolanlawson) [inline packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts:171]:
```suggestion
    const { isSlotted } = cxt;
```

Nitpick, but let's keep the naming consistent. 🙂

File: packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts [hunk 6]
```diff
@@ -199,6 +203,8 @@ export function getSlottedContent(
         lightSlotContent.length > 0 || scopedSlotContent.length > 0
     );
 
+    cxt.isSlotted = isNested;
```
Comment (nolanlawson) [inline packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts:206]:
```suggestion
    cxt.isSlotted = isSlotted;
```

**Review by nolanlawson — APPROVED** [review]
Minor nits, LGTM! :shipit:

**Comment by cardoso** [comment]
It seems `slot-not-at-top-level/advanced/lwcIf/shadow/index.js` broke after I committed the suggested changes. Trying to figure out why...

**Comment by cardoso** [comment]
I added `slot-not-at-top-level/advanced/lwcIf/shadow/index.js` back to expected failures. Not sure why it wasn't failing locally before, but it makes sense since the light counterpart is also an expected failure. This should be addressed in my follow-up PR.

**Pull request merged**